samza-core/src/main/java/org/apache/samza/operators/impl/WatermarkStates.java

```diff
@@ -0,0 +1,113 @@
+
+package org.apache.samza.operators.impl;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.samza.system.SystemStream;
+import org.apache.samza.system.SystemStreamPartition;
+import org.apache.samza.system.WatermarkMessage;
+
+import static org.apache.samza.operators.functions.WatermarkFunction.WATERMARK_NOT_EXIST;
+
+
+/**
+ * This class manages the state of the watermarks in a task. Internally it keeps track of the latest
+ * watermark timestamp from each upstream task, and use the min as the consolidated watermark time.
+ */
+class WatermarkStates {
+
+  private final static class WatermarkState {
+    private int expectedTotal = Integer.MAX_VALUE;
+    private final Map<String, Long> timestamps = new HashMap<>();
+    private long watermarkTime = WATERMARK_NOT_EXIST;
+
+    WatermarkState(int expectedTotal) {
+      this.expectedTotal = expectedTotal;
+    }
+
+    boolean update(long timestamp, String taskName) {
+      final long preWatermarkTime = watermarkTime;
+      if (taskName != null) {
+        timestamps.put(taskName, timestamp);
+      }
+
+      if (timestamps.size() == expectedTotal) {
+        Optional<Long> min = timestamps.values().stream().min(Long::compare);
+        watermarkTime = min.orElse(timestamp);
+      }
+
+      return preWatermarkTime != watermarkTime;
+    }
+
+    long getWatermarkTime() {
+      return watermarkTime;
+    }
+  }
+
+  private final Map<SystemStreamPartition, WatermarkState> watermarkStates;
+  private final Map<SystemStream, Long> watermarks = new HashMap<>();
+
+  WatermarkStates(Set<SystemStreamPartition> ssps, Map<SystemStream, Integer> producerTaskCounts) {
+    Map<SystemStreamPartition, WatermarkState> states = new HashMap<>();
+    ssps.forEach(ssp -> {
+        states.put(ssp, new WatermarkState(producerTaskCounts.getOrDefault(ssp.getSystemStream(), 0)));
+        watermarks.put(ssp.getSystemStream(), WATERMARK_NOT_EXIST);
+      });
+    this.watermarkStates = Collections.unmodifiableMap(states);
+  }
+
+  /**
+   * Update the state upon receiving a watermark message.
+   * @param watermarkMessage message of {@link WatermarkMessage}
+   * @param ssp system stream partition
+   * @return true iff the stream has a new watermark
+   */
+  boolean update(WatermarkMessage watermarkMessage, SystemStreamPartition ssp) {
+    WatermarkState state = watermarkStates.get(ssp);
+    boolean updated = state.update(watermarkMessage.getTimestamp(), watermarkMessage.getTaskName());
+
+    if (updated) {
+      long minTimestamp = watermarkStates.entrySet().stream()
+          .filter(entry -> entry.getKey().getSystemStream().equals(ssp.getSystemStream()))
+          .map(entry -> entry.getValue().getWatermarkTime())
+          .min(Long::compare)
+          .get();
+      Long curWatermark = watermarks.get(ssp.getSystemStream());
+      if (curWatermark == null || curWatermark < minTimestamp) {
```

Comment:
I believe the first condition would never hold in the first implementation. In that case it would seem clearer to assert that.

Reply:
That's correct. I added the assert.